### PR TITLE
Inspection warning about procedure coercion calls

### PR DESCRIPTION
**Pull request by MDoerner**
The new `ObjectWhereProcedureIsRequiredInspection` warns about places in which a _procedure coercion_ takes place, i.e. places in which an object is provided in a call statement. This leads to a call to the default member, provided it is a procedure or function or the default member cannot be bound at compile time. (Properties yield a compilation error, since they are not allowed in call statements.) This type of indirection is next to never really necessary and much more likely is caused by erroneously forgetting a member call.

This PR also provides the `ExpandDefaultMemberQuickFiz`, which does the obvious thing. Note that it will always expand the full default member chain for recursive default member reolutions. Moreover, it is deactivated for any such chain that cannot fully be bound at compile time. This is the first step towards fixing issue #2504; just the corresponding inspection is missing.

NOTE:
Because of the infrastructure introduced in PR #5164, this PR is based on it. Please do not merge before that PR. I will rebase this one once the other PR is merged.

### CODE REVIEW

**Comment by bclothier** [comment]
LGTM; just one suggestion - In test cases we consider only the scenario of `cls` (where it should be `cls.Foo`. However, would it not be more common to see the misuses in cases like `Bar(cls)`? Having an explicit test for that case would at least guarantee that the inspection is not influenced by the different syntax.

File: Rubberduck.CodeAnalysis/Inspections/ObjectWhereProcedureIsRequiredInspection.cs [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+    ///
+    /// Public Function Foo() As Long
+    /// Attibute VB_UserMemId = 0
+    ///     Foo = 42
+    /// End Function
+    ///
+    /// Module:
+    /// 
+    /// Public Sub DoSomething(ByVal arg As Class1)
+    ///     arg
```
Comment (retailcoder) [inline Rubberduck.CodeAnalysis/Inspections/ObjectWhereProcedureIsRequiredInspection.cs]: Would `Debug.Print arg` make a better / more realistic example?

File: Rubberduck.Resources/Inspections/InspectionInfo.resx [hunk 2]
```diff
@@ -412,4 +412,7 @@ If the parameter can be null, ignore this inspection result; passing a null valu
   <data name="UseOfUnboundBangNotationInspection" xml:space="preserve">
     <value>Bang notation, formally known as dictionary access expression, looks like it is strongly typed. However, it is actually a stringly-typed access to the parameterized default member of the object it is used on. This is especially misleading the default member cannot be determined at compile time.</value>
   </data>
+  <data name="ObjectWhereProcedureIsRequiredInspection" xml:space="preserve">
+    <value>Using an object with a default member in a place that requires a procedure leads to a call to the default member. This is most likely an error, but at least not obvious.</value>
```
Comment (retailcoder) [inline Rubberduck.Resources/Inspections/InspectionInfo.resx]:
```suggestion
    <value>Using an object with a default member in a place that requires a procedure leads to an implicit invocation of the default member. This is most likely an unintentional omission, and negatively affects readability.</value>
```

File: Rubberduck.CodeAnalysis/Inspections/ObjectWhereProcedureIsRequiredInspection.cs (hunk 1 above)
Comment (MDoerner) [reply Rubberduck.CodeAnalysis/Inspections/ObjectWhereProcedureIsRequiredInspection.cs]: No, then it is not a procedure coercion.

File: Rubberduck.Resources/Inspections/InspectionInfo.resx [hunk 3]
```diff
@@ -413,6 +413,6 @@ If the parameter can be null, ignore this inspection result; passing a null valu
     <value>Bang notation, formally known as dictionary access expression, looks like it is strongly typed. However, it is actually a stringly-typed access to the parameterized default member of the object it is used on. This is especially misleading the default member cannot be determined at compile time.</value>
   </data>
   <data name="ObjectWhereProcedureIsRequiredInspection" xml:space="preserve">
-    <value>Using an object with a default member in a place that requires a procedure leads to a call to the default member. This is most likely an error, but at least not obvious.</value>
+    <value>Using an object with a default member in a place that requires a procedure leads to an implicit invocation of the default member. This is most likely unintentional, at least negatively affects readability.</value>
```
Comment (bclothier) [inline Rubberduck.Resources/Inspections/InspectionInfo.resx]: The last sentence is fragmentary --- maybe: 

```
This is most likely unintentional and negatively affects readability.
```

File: Rubberduck.CodeAnalysis/Inspections/ObjectWhereProcedureIsRequiredInspection.cs [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+using Rubberduck.Parsing.VBA.DeclarationCaching;
+using Rubberduck.VBEditor;
+
+namespace Rubberduck.Inspections.Concrete
+{
+    /// <summary>
+    /// Identifies places in which an object is used but a procedure is required and a default member exists on the object.
+    /// </summary>
+    /// <why>
+    /// Providing an object in a place in which a procedure is required leads to a call to the objects default member.
```
Comment (retailcoder) [inline Rubberduck.CodeAnalysis/Inspections/ObjectWhereProcedureIsRequiredInspection.cs]:
```suggestion
    /// Providing an object where a procedure is required leads to an implicit call to the object's default member.
```

File: Rubberduck.CodeAnalysis/Inspections/ObjectWhereProcedureIsRequiredInspection.cs [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+    /// This behavior is not obvious and most likely an error.
```
Comment (retailcoder) [inline Rubberduck.CodeAnalysis/Inspections/ObjectWhereProcedureIsRequiredInspection.cs]:
```suggestion
    /// This behavior is not obvious, and most likely unintended.
```

**Pull request merged**